### PR TITLE
Added api/v2/search.py route.

### PR DESCRIPTION
**Pull request by p0psicles**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

Part of #6709 

This will introduce a number of search-related apiv2 endpoints.

GET: `http://localhost:8081/api/v2/search/tvdb276173` get a recent, running or queued searches for a specific show.

POST: `http://localhost:8081/api/v2/search/manual`
data: 
```
{
	"showslug": "tvdb289909",
	"episodes": [
	    "s03e03"
	],
	"options": {}
}
```

manual can be interchanged with:
* backlog
* daily
* failed

This api endpoint is currently not used. But will be when all PR's for #6709 have been merged.
More changes to the search route will be added with PR #6718 

### TODO:
- [ ] Fix dredd tests
- [ ] Fix season search (add the option)
```
season: [
        "s01",
         "s02",
         "s03",
     ],
```

### CODE REVIEW

File: medusa/server/api/v2/search.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+# coding=utf-8
+"""Request handler for statistics."""
+from __future__ import unicode_literals
+
+from collections import defaultdict
+from medusa import app
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: Correct import order:
```py
from __future__ import unicode_literals

from collections import defaultdict

from medusa import app
from medusa.search.manual import collect_episodes_from_search_thread
from medusa.search.queue import (
    BacklogQueueItem,
    FailedQueueItem,
    ForcedSearchQueueItem,
    # ManualSearchQueueItem,
)
from medusa.server.api.v2.base import BaseRequestHandler
from medusa.tv.episode import Episode, EpisodeNumber
from medusa.tv.series import Series, SeriesIdentifier

from tornado.escape import json_decode
```

File: medusa/server/api/v2/search.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            season_segments[episode.season].append(episode)
+
+        if not season_segments:
+            return self._not_found('No episodes passed to search for using the backlog search')
+
+        # Retrieve the search type option (episode vs season)
+        search_type = data['options'].get('type', 'episode')
+
+        for segment in season_segments.values():
+            ## Need this as it's part of a followup
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: `E266 too many leading '#' for block comment`

File: medusa/server/api/v2/search.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+class SearchHandler(BaseRequestHandler):
+    """Search queue request handler."""
+
+    #: resource name
+    name = 'search'
+    #: identifier
+    identifier = ('identifier', r'\w+')
+    #: path param
+    path_param = ('path_param', r'\w+')
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: If you don't plan on using `path_param` in the future, you should set it to `None` here.
For example:
https://github.com/pymedusa/Medusa/blob/6065fdc1590c3c885bf353ccd75cc3cac945f73b/medusa/server/api/v2/internal.py#L31-L32

File: medusa/server/api/v2/search.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    #: allowed HTTP methods
+    allowed_methods = ('GET', 'POST')
+
+    def get(self, identifier, path_param=None, *args, **kwargs):
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: Any reason for the `*args, **kwargs`?

File: medusa/server/api/v2/search.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        """Queue a backlog search for a range of episodes.
+
+        :param identifier:
+        :param path_param:
+        :type path_param: str
+        "tvdb1234" : [
+            "s01e01",
+            "s01e02",
+            "s03e03",
+        ]
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: Was this part of the docstring added intentionally?

File: medusa/server/api/v2/search.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+    def _search_backlog(self, data):
+        """Queue a backlog search for results for the provided episodes.
+
+        :param data:
+        :return:
+        """
+        statuses = {}
+
+        if all([not data.get('showslug'), not data.get('episodes')]):
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: I think it should be
```suggestion
        if all([not data.get('showSlug'), not data.get('episodes')]):
```

File: medusa/server/api/v2/search.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+            return self._bad_request('Triggering a backlog search failed')
+
+        if not data.get('showslug'):
+            return self._bad_request('For a backlog search you need to provide a showslug')
+
+        if not data.get('episodes'):
+            return self._bad_request('For a backlog search you need to provide a list of episodes')
+
+        identifier = SeriesIdentifier.from_slug(data['showslug'])
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        identifier = SeriesIdentifier.from_slug(data['showSlug'])
```

File: medusa/server/api/v2/search.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+        if all([not data.get('showslug'), not data.get('episodes')]):
+            # Trigger a full backlog search
+            if app.backlog_search_scheduler.forceRun():
+                return self._created()
+
+            return self._bad_request('Triggering a backlog search failed')
+
+        if not data.get('showslug'):
+            return self._bad_request('For a backlog search you need to provide a showslug')
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
            return self._bad_request('For a backlog search you need to provide a show slug')
```

File: medusa/server/api/v2/search.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        """
+        statuses = {}
+
+        if not data.get('showslug'):
+            return self._bad_request('For a backlog search you need to provide a showslug')
+
+        if not data.get('episodes'):
+            return self._bad_request('For a backlog search you need to provide a list of episodes')
+
+        identifier = SeriesIdentifier.from_slug(data['showslug'])
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        identifier = SeriesIdentifier.from_slug(data['showSlug'])
```

File: medusa/server/api/v2/search.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    def _search_failed(self, data):
+        """Queue a failed search.
+
+        :param data:
+        :return:
+        """
+        statuses = {}
+
+        if not data.get('showslug'):
+            return self._bad_request('For a backlog search you need to provide a showslug')
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
            return self._bad_request('For a backlog search you need to provide a show slug')
```

File: medusa/server/api/v2/search.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    def _search_manual(self, data):
+        """Queue a manual search for results for the provided episodes.
+
+        :param data:
+        :return:
+        """
+        statuses = {}
+
+        if not data.get('showslug'):
+            return self._bad_request('For a backlog search you need to provide a showslug')
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
            return self._bad_request('For a backlog search you need to provide a show slug')
```

File: medusa/server/api/v2/search.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+        if not data.get('episodes'):
+            return self._bad_request('For a backlog search you need to provide a list of episodes')
+
+        identifier = SeriesIdentifier.from_slug(data['showslug'])
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        identifier = SeriesIdentifier.from_slug(data['showSlug'])
```

File: medusa/server/api/v2/search.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+        statuses = {}
+
+        if all([not data.get('showslug'), not data.get('episodes')]):
+            # Trigger a full backlog search
+            if app.backlog_search_scheduler.forceRun():
+                return self._created()
+
+            return self._bad_request('Triggering a backlog search failed')
+
+        if not data.get('showslug'):
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        if not data.get('showSlug'):
```

File: medusa/server/api/v2/search.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+
+    def _search_failed(self, data):
+        """Queue a failed search.
+
+        :param data:
+        :return:
+        """
+        statuses = {}
+
+        if not data.get('showslug'):
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        if not data.get('showSlug'):
```

File: medusa/server/api/v2/search.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+    def _search_manual(self, data):
+        """Queue a manual search for results for the provided episodes.
+
+        :param data:
+        :return:
+        """
+        statuses = {}
+
+        if not data.get('showslug'):
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        if not data.get('showSlug'):
```

File: medusa/server/api/v2/search.py [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+            if not episode:
+                statuses[episode_slug] = {'status': 404}
+                continue
+
+            season_segments[episode.season].append(episode)
+
+        if not season_segments:
+            return self._not_found('No episodes passed to search for using the backlog search')
+
+        for segment in season_segments.values():
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]: I think this should use `itervalues` from `six`
```suggestion
        for segment in itervalues(season_segments):
```

File: medusa/server/api/v2/search.py [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        for segment in itervalues(season_segments):
```

File: medusa/server/api/v2/search.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+            season_segments[episode.season].append(episode)
+
+        if not season_segments:
+            return self._not_found('No episodes passed to search for using the backlog search')
+
+        # Retrieve the search type option (episode vs season)
+        search_type = data['options'].get('type', 'episode')
+
+        for segment in season_segments.values():
```
Comment (sharkykh) [inline medusa/server/api/v2/search.py]:
```suggestion
        for segment in itervalues(season_segments):
```

**Comment by p0psicles** [comment]
@sharkykh tnx for the review. Moet that any of thuis code Will change allot. But i'm still trying to get An overview of all the pr's i need. So for consider alle these pr's as drafts. Some may even be dropped.

File: medusa/server/api/v2/search.py (hunk 4 above)
Comment (p0psicles) [reply medusa/server/api/v2/search.py]: hmm yeah if I don't intend to use `path_param`. I can now use something like: `def get(self, identifier, *args, **kwargs):`

**Comment by p0psicles** [comment]
@sharkykh i'm continuing with this one

File: medusa/server/api/v2/search.py (hunk 5 above)
Comment (p0psicles) [reply medusa/server/api/v2/search.py]: yes. I've cleaned it up a little by adding it with the :example: section.

**Comment by p0psicles** [comment]
I'm also working on the API-description. I'll commit it in the next few days

File: medusa/server/api/v2/search.py (hunk 4 above)
Comment (sharkykh) [reply medusa/server/api/v2/search.py]: I believe the APIv2 base handler is set up in a way that it doesn't push extra arguments.
So you could remove this - most endpoints don't have `*args, **kwargs`.

**Comment by sharkykh** [comment]
(just in case it's hidden) https://github.com/pymedusa/Medusa/pull/6716#discussion_r298246910

File: medusa/server/api/v2/search.py (hunk 4 above)
Comment (p0psicles) [reply medusa/server/api/v2/search.py]: this should be fixed

**Comment by p0psicles** [comment]
@medariox the concept of the "season" search only applies to the `/apiv2/search/manual` right? As for the backlog search, it's already handled by checking if all episodes for a season are passed in the segment.

**Comment by medariox** [comment]
@p0psicles 
I think so. I believe that using the API for both would be better, as it's more consistent.

File: dredd/api-description.yml [hunk 19]
```diff
@@ -936,40 +926,75 @@ paths:
...
       description: Queue a forced daily search
       responses:
-        201:
-          description: "Queue a new daily search"
+        202:
+          description: Queue a new daily search
         400:
           $ref: '#/responses/error'
+          x-disabled: true
+  /search/{seriesid}:
```
Comment (p0psicles) [inline dredd/api-description.yml]: This is duplicate now

Comment (p0psicles) [reply dredd/api-description.yml]: Ow you moved it. Never mind

File: medusa/server/api/v2/search.py [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+        series = Series.find_by_identifier(identifier)
+        if not series:
+            return self._not_found('Series not found')
+
+        episode_segments = self._get_episode_segments(series, data)
+
+        # If a season is passed, we transform it to a list of episode objects. And merge it with the episode_segment.
+        # This because the backlog search has it's own logic for searching per episode or season packs. And falling back
+        # between them, if configured.
+        if data.get('season') and len(data.get('season')) > 0:
```
Comment (medariox) [inline medusa/server/api/v2/search.py]: I'm a bit confused about `len(data.get('season')) > 0`. What is it for?

Comment (p0psicles) [reply medusa/server/api/v2/search.py]: I think it was to check for to have at least one season slug. But in hindsight, I don't think it's required.

**Pull request merged**